api/include/opentelemetry/trace/span.h

```diff
@@ -1,38 +1,112 @@
+  kProducer,
+  kConsumer,
+};
 /**
- * StartSpanOptions provides options to set properties of span at the time of starting a new span.
+ * StartSpanOptions provides options to set properties of a Span at the time of its creation
  */
 struct StartSpanOptions
 {
-  // Sets the start time of a span.
+  // Optionally sets the start time of a Span.
   //
-  // If the start time of a span is set, timestamps from both the system clock and steady clock
-  // should be provided.
+  // If the start time of a Span is set, timestamps from both the system clock and steady clock
+  // must be provided.
   //
-  // Timestamps from the steady clock can be used to most accurately measure a span's
-  // duration, while timestamps from the system clock can be used to most accurately place a span's
-  // time point relative to other spans collected across a distributed system.
+  // Timestamps from the steady clock can be used to most accurately measure a Span's
+  // duration, while timestamps from the system clock can be used to most accurately place a Span's
+  // time point relative to other Spans collected across a distributed system.
   core::SystemTimestamp start_system_time;
   core::SteadyTimestamp start_steady_time;
+
+  // TODO:
+  // Span(Context?) parent;
+  // SpanContext remote_parent;
+  // Links
+  // Attributes
+  SpanKind kind = SpanKind::kInternal;
 };
 
 class Tracer;
 
 /**
- * A span represents a single operation within a trace.
+ * A Span represents a single operation within a Trace.
+ *
+ * Spans *must* be ended by calling End(). Merely deleting a Span does not end
+ * it.
  */
 class Span
 {
 public:
+  // Note that Spans should be created using the Tracer class. Please refer to
+  // tracer.h for documentation.
+  Span()          = default;
   virtual ~Span() = default;
 
+  // Not copiable or movable.
+  Span(const Span &) = delete;
+  Span(Span &&)      = delete;
+  Span &operator=(const Span &) = delete;
+  Span &operator=(Span &&) = delete;
+
+  // TODO
+  // Sets an attribute on the Span. If the Span previously contained a mapping for
+  // the key, the old value is replaced.
+  //
+  // If an empty string is used as the value, the attribute will be silently
+  // dropped. Note: this behavior could change in the future.
+  // virtual void SetAttribute(nostd::string_view key, AttributeValue&& value) = 0;
+
+  // Adds an event to the Span.
+  virtual void AddEvent(nostd::string_view name) noexcept = 0;
+
+  // Adds an event to the Span, with a custom timestamp.
+  virtual void AddEvent(nostd::string_view name, core::SystemTimestamp timestamp) noexcept = 0;
+  virtual void AddEvent(nostd::string_view name, core::SteadyTimestamp timestamp) noexcept = 0;
+
+  // TODO
+  // Adds an event to the Span, with a custom timestamp, and attributes.
+  // virtual void AddEvent(nostd::string_view name, core::SteadyTimestamp
+  // timestamp, nostd::span<std::pair<nostd::string_view name, AttributeValue
+  // value>> attributes) noexcept = 0;
```

Comment:
This should be `nostd::span<const std::pair<nostd::string_view, AttributeValue>>`  since it doesn't modify the std::pair's (and otherwise it won't work with const containers)

Reply:
Done.

Is `std::pair` allowed, ABI-wise, or do we need to add a `nostd::pair` class?